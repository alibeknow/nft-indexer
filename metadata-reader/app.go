package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
