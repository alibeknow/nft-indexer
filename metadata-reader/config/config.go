package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
