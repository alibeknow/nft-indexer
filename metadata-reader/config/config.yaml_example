server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
