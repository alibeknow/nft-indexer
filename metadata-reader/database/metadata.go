package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
