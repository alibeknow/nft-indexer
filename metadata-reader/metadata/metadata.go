package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
