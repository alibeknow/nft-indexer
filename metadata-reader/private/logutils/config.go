package logutils

type Config struct {
	Level string
}
