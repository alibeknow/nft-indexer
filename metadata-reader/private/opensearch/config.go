package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
