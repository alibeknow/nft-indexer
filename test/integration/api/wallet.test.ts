import { HttpServer, INestApplication } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import { LoggerModule } from 'nestjs-pino';
import request from 'supertest';
import { IApiConfig, getApiConfig } from '@api/app.config';
import { WalletModule } from '@api/wallet';
import { DB, DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { Blockchain } from '@shared/blockchain';
import { ExplorerEvent, ExplorersClient } from '@shared/explorers-client';
import { Metadata, MetadataRepoDB } from '@shared/metadata';
import { getNestApp } from '../../helpers/app';

import {
  explorersApiErc1155EventsFixture,
  explorersApiErc721And1155EventsFixture,
  explorersApiErc721EventsFixture,
  metadataDbFixture,
} from './fixtures';

jest.setTimeout(50000);

const TRANSFER_EVENT_ID = '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef';
const TRANSFER_BATCH_EVENT_ID = '0x4a39dc06d4c0dbc64b70af90fd698a233a518aa5d07e595d983b8c0526c8f7fb';

async function seedMetadataRepo(db: DBClass): Promise<(Metadata | null)[]> {
  const metadataRepo = new MetadataRepoDB(db);

  return Promise.all(metadataDbFixture.map(
    ({ _id, metadata, type }) => metadataRepo.save(_id, metadata, type)),
  );
}

describe('Wallet endpoints', () => {
  let app: INestApplication;
  let httpServer: HttpServer;
  let db: DBClass;
  let explorersClient: ExplorersClient<IApiConfig>;

  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `api-wallet-test-${dbConfig.db}` });
    await db.open();

    const moduleRef = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          load: [ getApiConfig ],
          isGlobal: true,
        }),
        WalletModule,
        LoggerModule.forRoot({ pinoHttp: { enabled: false } }),
      ],
    })
      .overrideProvider(DB)
      .useValue(db)
      .compile();

    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();
    explorersClient = app.get<ExplorersClient<IApiConfig>>(ExplorersClient<IApiConfig>);
  });

  afterEach(async () => {
    await db.metadata().deleteMany({});
  });

  afterAll(async () => {
    await db.close(true);
    await app.close();
  });

  describe('[GET /api/v0/:chainName/wallet/:address/nfts]', () => {
    it('should return validation error if wrong address is passed', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/fake-addrs/nfts')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return validation error if wrong chain name is passed', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v0/fake-chain-name/wallet/0x7732920c57ed119f5000f7bd3a34a954b7932ed2/nfts')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return validation error if incorrect query params were passed', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0x7732920c57ed119f5000f7bd3a34a954b7932ed2/nfts?page=0&limit=some-limit')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return empty array if explorers api returns no events', async () => {
      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async () => {
        return [];
      });

      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0x7732920c57ed119f5000f7bd3a34a954b7932ed2/nfts')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct result for erc-721 tokens', async () => {
      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async (
        chainName: Blockchain, walletAddress: string, eventSignature: string,
      ) => {
        if(eventSignature === TRANSFER_EVENT_ID) {
          return explorersApiErc721EventsFixture as unknown as ExplorerEvent[];
        }

        return [];
      });

      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0x47ba73787ea555cafbf1c4472b03471c73e46e76/nfts')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct result for erc-1155 tokens', async () => {
      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async (
        chainName: Blockchain, walletAddress: string, eventSignature: string,
      ) => {
        if(eventSignature === TRANSFER_BATCH_EVENT_ID) {
          return explorersApiErc1155EventsFixture as unknown as ExplorerEvent[];
        }

        return [];
      });

      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0x3ee13342db4bc021233f9e9f5c83d7e0f0995ae3/nfts')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct result for erc-1155 and erc-721 tokens at the same time', async () => {
      await seedMetadataRepo(db);

      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async (
        chainName: Blockchain, walletAddress: string, eventSignature: string,
      ) => {
        if(eventSignature === TRANSFER_BATCH_EVENT_ID) {
          return explorersApiErc721And1155EventsFixture as unknown as ExplorerEvent[];
        }

        return [];
      });

      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0x14e27b3f72721aabe20e4ec713e65e49422012da/nfts')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct results based on page and limit query params', async () => {
      await seedMetadataRepo(db);

      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async (
        chainName: Blockchain, walletAddress: string, eventSignature: string,
      ) => {
        if(eventSignature === TRANSFER_BATCH_EVENT_ID) {
          return explorersApiErc721And1155EventsFixture as unknown as ExplorerEvent[];
        }

        return [];
      });

      const actualResult1 = await request(httpServer)
        .get('/api/v0/eth/wallet/0x14e27b3f72721aabe20e4ec713e65e49422012da/nfts?page=2&limit=4')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v0/eth/wallet/0x14e27b3f72721aabe20e4ec713e65e49422012da/nfts?page=1&limit=3')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body.result).toHaveLength(4);
      expect(actualResult1.body).toMatchSnapshot();

      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body.result).toHaveLength(3);
      expect(actualResult2.body).toMatchSnapshot();
    });

    it('should return correct results based on contractAddress query param', async () => {
      await seedMetadataRepo(db);

      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async (
        chainName: Blockchain, walletAddress: string, eventSignature: string,
      ) => {
        if(eventSignature === TRANSFER_BATCH_EVENT_ID) {
          return explorersApiErc721And1155EventsFixture as unknown as ExplorerEvent[];
        }

        return [];
      });

      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0x14e27b3f72721aabe20e4ec713e65e49422012da/nfts?contractAddress=0x5090AF9eE4c65f250706972089445C9D78e31274')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body.result).toHaveLength(2);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct results based on non-checksum formatted contractAddress query param', async () => {
      await seedMetadataRepo(db);

      jest.spyOn(explorersClient, 'fetchWalletEvents').mockImplementation(async (
        chainName: Blockchain, walletAddress: string, eventSignature: string,
      ) => {
        if(eventSignature === TRANSFER_BATCH_EVENT_ID) {
          return explorersApiErc721And1155EventsFixture as unknown as ExplorerEvent[];
        }

        return [];
      });

      const actualResult = await request(httpServer)
        .get('/api/v0/eth/wallet/0X14e27b3f72721aabe20e4ec713e65e49422012da/nfts?contractAddress=0X5090af9eE4C65F250706972089445c9D78e31274')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body.result).toHaveLength(2);
      expect(actualResult.body).toMatchSnapshot();
    });
  });
});
