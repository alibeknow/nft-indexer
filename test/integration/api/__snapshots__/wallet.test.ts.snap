// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct result for erc-721 tokens 1`] = `
{
  "result": [
    {
      "contract": "0xd532b88607B1877Fe20c181CBa2550E3BbD6B31C",
      "tokenId": "0x1edc",
    },
    {
      "contract": "0x6fA9F4B50E2950a8137A76649193816FB29DAd2c",
      "tokenId": "0x06de",
    },
    {
      "contract": "0x6fA9F4B50E2950a8137A76649193816FB29DAd2c",
      "tokenId": "0x2e1b",
    },
    {
      "contract": "0xB318d1B6805D2F8Ec0e95e2600ad60b4725827C7",
      "tokenId": "0x75",
    },
    {
      "contract": "0xB318d1B6805D2F8Ec0e95e2600ad60b4725827C7",
      "tokenId": "0x0878",
    },
    {
      "contract": "0x753F10598C026E73182CA74ED33DE05974b9f083",
      "tokenId": "0x3147",
    },
    {
      "contract": "0x753F10598C026E73182CA74ED33DE05974b9f083",
      "tokenId": "0x3148",
    },
    {
      "contract": "0x753F10598C026E73182CA74ED33DE05974b9f083",
      "tokenId": "0x3149",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct result for erc-1155 and erc-721 tokens at the same time 1`] = `
{
  "result": [
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "background-96",
        },
        {
          "trait_type": "Flare",
          "value": "Light red",
        },
        {
          "trait_type": "Clothes",
          "value": "Striped vest",
        },
        {
          "trait_type": "Ears",
          "value": "Normal",
        },
        {
          "trait_type": "Mouth",
          "value": "Tongue out",
        },
        {
          "trait_type": "Eyes",
          "value": "Laser eye",
        },
        {
          "trait_type": "Top",
          "value": "Party",
        },
      ],
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmRYe4DGMqBUquW6yzHFRFBNMU794ygRrxvFBwLJ6xfArG/1088.png",
        },
      ],
      "contract": "0xEb59F3e65190C07379252E1910656d75C753C9A9",
      "description": "Rich ape #1088 from unique collection, find out more at https://www.richapes.club",
      "id": 1088,
      "image": "ipfs://QmRYe4DGMqBUquW6yzHFRFBNMU794ygRrxvFBwLJ6xfArG/1088.png",
      "tokenId": "0x00",
    },
    {
      "attributes": [
        {
          "trait_type": "Fur",
          "value": "Stone",
        },
        {
          "trait_type": "Mouth",
          "value": "Tree",
        },
        {
          "trait_type": "Eyes",
          "value": "Alien Eyes",
        },
        {
          "trait_type": "Clothes",
          "value": "Live Fire Coat",
        },
        {
          "trait_type": "Head",
          "value": "Army Helmet",
        },
      ],
      "content": [
        {
          "type": "image",
          "url": "https://teenapegang.s3.amazonaws.com/mutants/0806.png",
        },
      ],
      "contract": "0x90f94E450Ac1c3E1D2C97A72E778E783C558Bf3E",
      "description": "",
      "image": "https://teenapegang.s3.amazonaws.com/mutants/0806.png",
      "name": "Teen Ape Gang Mutant #832",
      "tokenId": "0x02",
    },
    {
      "attributes": [
        {
          "trait_type": "Fur",
          "value": "Aquarium",
        },
        {
          "trait_type": "Mouth",
          "value": "Tree",
        },
        {
          "trait_type": "Eyes",
          "value": "Eyes Melting",
        },
        {
          "trait_type": "Head",
          "value": "Bleeding Heli Hat",
        },
      ],
      "content": [
        {
          "type": "image",
          "url": "https://teenapegang.s3.amazonaws.com/mutants/0002.png",
        },
      ],
      "contract": "0x3F95b55b591e188D845e1d68531555322C8A5507",
      "description": "",
      "image": "https://teenapegang.s3.amazonaws.com/mutants/0002.png",
      "name": "Teen Ape Gang Mutant #833",
      "tokenId": "0x0b67",
    },
    {
      "contract": "0x735F47aD87FEF27De6179d8c4b7dCdAC5DfC077F",
      "tokenId": "0x02e1",
    },
    {
      "contract": "0xFf003B436C441ea8377Dd3756B0E42E549D2704c",
      "tokenId": "0x035e",
    },
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "Orange",
        },
        {
          "trait_type": "Body",
          "value": "Body",
        },
        {
          "trait_type": "Beak",
          "value": "Upset",
        },
        {
          "trait_type": "Eyes",
          "value": "Scared",
        },
        {
          "trait_type": "Eyebrows",
          "value": "Default",
        },
        {
          "trait_type": "Trousers",
          "value": "Pink",
        },
        {
          "trait_type": "Shirt",
          "value": "None",
        },
        {
          "trait_type": "Feet",
          "value": "Smart Shoes",
        },
        {
          "trait_type": "Hat",
          "value": "Green Bucket Hat",
        },
        {
          "trait_type": "Item",
          "value": "Pancakes",
        },
      ],
      "compiler": "Engine",
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
        },
      ],
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "date": 1643342687198,
      "description": "The star from AlphaBetty Books first publication.",
      "dna": "8763393910575a1668ec6696617204632dd80db9",
      "edition": 9745,
      "image": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
      "name": "The Seagull finds his talent #9745",
      "tokenId": "0x0a29",
    },
    {
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "tokenId": "0x0a2a",
    },
    {
      "contract": "0x81b491EA8c26AffFE00Dd27517Dcd838Bf5097BE",
      "tokenId": "0x0918",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct result for erc-1155 tokens 1`] = `
{
  "result": [
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x19",
    },
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x12",
    },
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x0d",
    },
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x01",
    },
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x09",
    },
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x07",
    },
    {
      "contract": "0x5c891d76584b46bC7F1E700169a76569Bb77d2Db",
      "tokenId": "0x11",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct results based on contractAddress query param 1`] = `
{
  "result": [
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "Orange",
        },
        {
          "trait_type": "Body",
          "value": "Body",
        },
        {
          "trait_type": "Beak",
          "value": "Upset",
        },
        {
          "trait_type": "Eyes",
          "value": "Scared",
        },
        {
          "trait_type": "Eyebrows",
          "value": "Default",
        },
        {
          "trait_type": "Trousers",
          "value": "Pink",
        },
        {
          "trait_type": "Shirt",
          "value": "None",
        },
        {
          "trait_type": "Feet",
          "value": "Smart Shoes",
        },
        {
          "trait_type": "Hat",
          "value": "Green Bucket Hat",
        },
        {
          "trait_type": "Item",
          "value": "Pancakes",
        },
      ],
      "compiler": "Engine",
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
        },
      ],
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "date": 1643342687198,
      "description": "The star from AlphaBetty Books first publication.",
      "dna": "8763393910575a1668ec6696617204632dd80db9",
      "edition": 9745,
      "image": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
      "name": "The Seagull finds his talent #9745",
      "tokenId": "0x0a29",
    },
    {
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "tokenId": "0x0a2a",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct results based on non-checksum formatted contractAddress query param 1`] = `
{
  "result": [
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "Orange",
        },
        {
          "trait_type": "Body",
          "value": "Body",
        },
        {
          "trait_type": "Beak",
          "value": "Upset",
        },
        {
          "trait_type": "Eyes",
          "value": "Scared",
        },
        {
          "trait_type": "Eyebrows",
          "value": "Default",
        },
        {
          "trait_type": "Trousers",
          "value": "Pink",
        },
        {
          "trait_type": "Shirt",
          "value": "None",
        },
        {
          "trait_type": "Feet",
          "value": "Smart Shoes",
        },
        {
          "trait_type": "Hat",
          "value": "Green Bucket Hat",
        },
        {
          "trait_type": "Item",
          "value": "Pancakes",
        },
      ],
      "compiler": "Engine",
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
        },
      ],
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "date": 1643342687198,
      "description": "The star from AlphaBetty Books first publication.",
      "dna": "8763393910575a1668ec6696617204632dd80db9",
      "edition": 9745,
      "image": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
      "name": "The Seagull finds his talent #9745",
      "tokenId": "0x0a29",
    },
    {
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "tokenId": "0x0a2a",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct results based on page and limit query params 1`] = `
{
  "result": [
    {
      "contract": "0xFf003B436C441ea8377Dd3756B0E42E549D2704c",
      "tokenId": "0x035e",
    },
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "Orange",
        },
        {
          "trait_type": "Body",
          "value": "Body",
        },
        {
          "trait_type": "Beak",
          "value": "Upset",
        },
        {
          "trait_type": "Eyes",
          "value": "Scared",
        },
        {
          "trait_type": "Eyebrows",
          "value": "Default",
        },
        {
          "trait_type": "Trousers",
          "value": "Pink",
        },
        {
          "trait_type": "Shirt",
          "value": "None",
        },
        {
          "trait_type": "Feet",
          "value": "Smart Shoes",
        },
        {
          "trait_type": "Hat",
          "value": "Green Bucket Hat",
        },
        {
          "trait_type": "Item",
          "value": "Pancakes",
        },
      ],
      "compiler": "Engine",
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
        },
      ],
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "date": 1643342687198,
      "description": "The star from AlphaBetty Books first publication.",
      "dna": "8763393910575a1668ec6696617204632dd80db9",
      "edition": 9745,
      "image": "ipfs://QmanhhXvqBhBKbgxyQEYdmURaJjEeKPfDFZbjThC5in6dx/9745",
      "name": "The Seagull finds his talent #9745",
      "tokenId": "0x0a29",
    },
    {
      "contract": "0x5090AF9eE4c65f250706972089445C9D78e31274",
      "tokenId": "0x0a2a",
    },
    {
      "contract": "0x81b491EA8c26AffFE00Dd27517Dcd838Bf5097BE",
      "tokenId": "0x0918",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return correct results based on page and limit query params 2`] = `
{
  "result": [
    {
      "attributes": [
        {
          "trait_type": "Background",
          "value": "background-96",
        },
        {
          "trait_type": "Flare",
          "value": "Light red",
        },
        {
          "trait_type": "Clothes",
          "value": "Striped vest",
        },
        {
          "trait_type": "Ears",
          "value": "Normal",
        },
        {
          "trait_type": "Mouth",
          "value": "Tongue out",
        },
        {
          "trait_type": "Eyes",
          "value": "Laser eye",
        },
        {
          "trait_type": "Top",
          "value": "Party",
        },
      ],
      "content": [
        {
          "type": "image",
          "url": "ipfs://QmRYe4DGMqBUquW6yzHFRFBNMU794ygRrxvFBwLJ6xfArG/1088.png",
        },
      ],
      "contract": "0xEb59F3e65190C07379252E1910656d75C753C9A9",
      "description": "Rich ape #1088 from unique collection, find out more at https://www.richapes.club",
      "id": 1088,
      "image": "ipfs://QmRYe4DGMqBUquW6yzHFRFBNMU794ygRrxvFBwLJ6xfArG/1088.png",
      "tokenId": "0x00",
    },
    {
      "attributes": [
        {
          "trait_type": "Fur",
          "value": "Stone",
        },
        {
          "trait_type": "Mouth",
          "value": "Tree",
        },
        {
          "trait_type": "Eyes",
          "value": "Alien Eyes",
        },
        {
          "trait_type": "Clothes",
          "value": "Live Fire Coat",
        },
        {
          "trait_type": "Head",
          "value": "Army Helmet",
        },
      ],
      "content": [
        {
          "type": "image",
          "url": "https://teenapegang.s3.amazonaws.com/mutants/0806.png",
        },
      ],
      "contract": "0x90f94E450Ac1c3E1D2C97A72E778E783C558Bf3E",
      "description": "",
      "image": "https://teenapegang.s3.amazonaws.com/mutants/0806.png",
      "name": "Teen Ape Gang Mutant #832",
      "tokenId": "0x02",
    },
    {
      "attributes": [
        {
          "trait_type": "Fur",
          "value": "Aquarium",
        },
        {
          "trait_type": "Mouth",
          "value": "Tree",
        },
        {
          "trait_type": "Eyes",
          "value": "Eyes Melting",
        },
        {
          "trait_type": "Head",
          "value": "Bleeding Heli Hat",
        },
      ],
      "content": [
        {
          "type": "image",
          "url": "https://teenapegang.s3.amazonaws.com/mutants/0002.png",
        },
      ],
      "contract": "0x3F95b55b591e188D845e1d68531555322C8A5507",
      "description": "",
      "image": "https://teenapegang.s3.amazonaws.com/mutants/0002.png",
      "name": "Teen Ape Gang Mutant #833",
      "tokenId": "0x0b67",
    },
  ],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return empty array if explorers api returns no events 1`] = `
{
  "result": [],
  "statusCode": 200,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return validation error if incorrect query params were passed 1`] = `
{
  "error": "Bad Request",
  "message": [
    "limit must be a positive number",
    "limit must be an integer number",
    "page must be a positive number",
  ],
  "statusCode": 400,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return validation error if wrong address is passed 1`] = `
{
  "error": "Bad Request",
  "message": [
    "address must be an Ethereum address",
  ],
  "statusCode": 400,
}
`;

exports[`Wallet endpoints [GET /api/v0/:chainName/wallet/:address/nfts] should return validation error if wrong chain name is passed 1`] = `
{
  "error": "Bad Request",
  "message": [
    "chainName must be a valid enum value",
  ],
  "statusCode": 400,
}
`;
